### ISSUE TITLE
Implement robust normalization options with continuous outlier indicators

### ISSUE THREAD

**Issue by DM-Berger** [stfxecutables/df-analyze#8]
**Summary**: The current robust normalization scheme destroys information in outliers and will poorly handle features with power law or exponential distributions. Some simple indicator schemes may fix this. 

## Indicator Approach for Clipping

When imputing NaNs, it is not uncommon to add a binary indicator variable that indicates a NaN was present. There is no reason we could not **add an indicator feature for values that were clipped**. This indicator could be:

- binary i.e. in {0, 1}, where 1 indicates "was clipped"
- scaled continuous 
  - with `x`, `a_min`, and `a_max` as above, after median centering:
    ```python
    x = np.copy(x)
    x[(x > a_min) & (x < a_max)] = 0
    x_indicator = MinMaxScaler().fit_transform(x)
    ```
- scaled log continuous
  - with `x`, `a_min`, and `a_max` as above, after median centering:
    ```python
    x = np.copy(x)
    x[(x > a_min) & (x < a_max)] = 0
    pos, neg = x > 0, x < 0
    x[pos] = np.log(x[pos] + 1)
    x[neg] = -np.log(np.abs(x[neg]) + 1)    
    x_indicator = MinMaxScaler().fit_transform(x)
    ```

Theoretically, if we have a way to detect exponentially or power-distributed variables, then the above normalization scheme could be extended to adding `n` bins for each clipping range, e.g. if `r = max(x) / min(x)`, then split `r` into `n` bins, and use the "scaled log continuous" method for each, but where the `a_min` and `a_max` correspond to each bin edge. This would make it so we have `n` different "zooms" or scaled views of the power-law / exponential, and would allow a learner to treat the variable differently at different scales, without losing any information at each scale. 

## Current Approach

Currently a modified form of robust normalization is used by default. With this robust method, the data is median-centred, and then the 5th and 95th data percentiles ("robust min/max") are calculated for each feature. The absolute difference between these percentiles is taken as the "robust range" of the data, and then values more than twice the robust range from these robust extremes are considered outliers, and clipped to the robust min/max, and then the data are min-max normalized. I.e. in code:

```python
import numpy as np
from sklearn.preprocessing import MinMaxScaler

def robust_normalize(x: np.ndarray) -> np.ndarray:
    x         -= np.median(x)  # x is our 1D feature
    rmin, rmax = np.percentile(x, [5, 95])
    rrange     = rmax - rmin
    a_max      = rmax + 2 * rrange
    a_min      = rmin - 2 * rrange
    x_clip     = np.clip(x, a_min, a_max)
    x_norm     = MinMaxScaler().fit_transform(x_clip)
    return x_norm
```

### Reasoning of Current Approach

- a sample is considered an outlier if it is two (robust) ranges beyond the (robust) extremes of the data
- it is better to clip and lose information in a few outliers rather than:
  - scale using the robust extrema, but have un-clipped values still be many orders of magnitude larger and act as potential anchor or influence points
  - scale non-robustly, and have a few outliers shrink all other sample feature values to near zero
- a factor of 2 (twice the range) from simulations of data with extreme outliers tends to result in distributions with 90% of the data being in [0, 0.5] or so, and outliers being around 1.0 (i.e. a single outlier will only be twice the maximum magnitude of normal features)
- for data without outliers, the factor of 2 seems to result in minimal or no observable change to the feature distribution

### Limitations of Current Approach

However, this method clearly:

1. destroys information in the outliers
2. cannot handle features distributed according to a power law or exponentially
